### PR TITLE
Fix/permissions

### PR DESCRIPTION
**Pull request by ShashiUnikon**
# Submit a pull request

## CLA

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [x] The code changes follow best practices
- [x] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: packages/stream_chat_flutter/lib/src/attachment/file_attachment.dart [hunk 1]
```diff
@@ -106,72 +112,131 @@ class _StreamFileAttachmentState extends State<StreamFileAttachment> {
...
+                  if (!isMyMessage) {
+                    Fluttertoast.showToast(msg: 'File not downloaded yet');
+                    return;
+                  }
+
+                  // Download the file
+                  FileDownloaderUtils.downloadFile2(
+                          url: assetUrl!,
+                          title: title!,
+                          messageId: widget.message.id)
```
Comment (DhruvinUnikon) [inline packages/stream_chat_flutter/lib/src/attachment/file_attachment.dart]: can we use async await here?

File: packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      }
+
+      final filePath = '${saveDir?.path}/$messageId/$title';
+
+      // Download the file
+      final dio = Dio();
+      await dio.download(url, filePath, onReceiveProgress: (received, total) {
+        if (total != -1) {
+          print(
+              'Download Progress: ${(received / total * 100).toStringAsFixed(0)}%');
```
Comment (DhruvinUnikon) [inline packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: we have to create ui for this.

File: packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        }
+      });
+
+      // Validate file existence
+      final file = File(filePath);
+      if (file.existsSync()) {
+        print('File downloaded successfully: $filePath');
```
Comment (DhruvinUnikon) [inline packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: print statements.

File: packages/stream_chat_flutter/lib/src/gallery/gallery_footer.dart [hunk 4]
```diff
@@ -81,50 +81,50 @@ class _StreamGalleryFooterState extends State<StreamGalleryFooter> {
...
+              //         [XFile(filePath)],
+              //         sharePositionOrigin: Rect.fromLTWH(
+              //           position.dx,
+              //           position.dy,
+              //           size?.width ?? 50,
+              //           (size?.height ?? 2) / 2,
+              //         ),
+              //       );
+              //     },
+              //   ),
```
Comment (DhruvinUnikon) [inline packages/stream_chat_flutter/lib/src/gallery/gallery_footer.dart:127]: remove if not required.

File: packages/stream_chat_flutter/pubspec.yaml [hunk 5]
```diff
@@ -50,6 +50,7 @@ dependencies:
   audio_waveforms: ^1.0.5
   open_file: ^3.5.9
   fluttertoast: ^8.2.8
+  permission_handler: any
```
Comment (DhruvinUnikon) [inline packages/stream_chat_flutter/pubspec.yaml:53]: better if can reference the particular reference here.

**Review by DhruvinUnikon — CHANGES_REQUESTED** [review]

**Comment by DhruvinUnikon** [comment]
Just a few lines away.
Great work!

File: packages/stream_chat_flutter/lib/src/gallery/gallery_footer.dart (hunk 4 above)
Comment (ShashiUnikon) [reply packages/stream_chat_flutter/lib/src/gallery/gallery_footer.dart:127]: This is stream implementation. As of now we have to hide the share button later we will uncomment this.

File: packages/stream_chat_flutter/lib/src/attachment/file_attachment.dart (hunk 1 above)
Comment (ShashiUnikon) [reply packages/stream_chat_flutter/lib/src/attachment/file_attachment.dart]: Sure sir!

Comment (ShashiUnikon) [reply packages/stream_chat_flutter/lib/src/attachment/file_attachment.dart]: Resolved

File: packages/stream_chat_flutter/pubspec.yaml (hunk 5 above)
Comment (ShashiUnikon) [reply packages/stream_chat_flutter/pubspec.yaml:53]: I have made it due to the packages compatibility

File: packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart (hunk 3 above)
Comment (ShashiUnikon) [reply packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: Removed

File: packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart [hunk 6]
```diff
@@ -37,19 +40,68 @@ class FileDownloaderUtils {
...
       Fluttertoast.showToast(msg: 'Failed to download file.');
       return null;
     }
   }
 
+  /// Get folder uri from shared preferences
+  static Future<String?> _getFolderUriFromSharedPreferences() async {
+    // Get the folderUri from persistent storage (e.g., SharedPreferences or Hive)
+    final pref = await SharedPreferences.getInstance();
+    final folderUri = pref.getString('folderUri');
```
Comment (ShashiUnikon) [inline packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: Make it a constant

File: packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart [hunk 7]
```diff
@@ -37,19 +40,68 @@ class FileDownloaderUtils {
...
+          data: File(sourceFile.path).readAsBytesSync(),
+          fileName: fileName);
+      return;
+    }
+
+    if (!await FlutterFileDialog.isPickDirectorySupported()) {
+      Fluttertoast.showToast(
+          msg: 'This feature is not supported on this device');
+      return;
+    }
```
Comment (ShashiUnikon) [inline packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: Move it at the top

Comment (ShashiUnikon) [reply packages/stream_chat_flutter/lib/src/file_downloader/file_downloader_utils.dart]: DOne

**Review by DhruvinUnikon — APPROVED** [review]
Great work!

**Pull request merged**